pkgs/pkgs-lib/formats/libconfig/test/comprehensive.nix

```diff
@@ -0,0 +1,50 @@
+{ lib, formats, stdenvNoCC, ... }:
+let
+  libconfig = formats.libconfig { };
+
+  include_expr = {
+    val = 1;
+  };
+
+  include_file = libconfig.generate "libconfig-test-include" include_expr;
+
+  expression = {
+    simple_top_level_attr = "1.0";
+    nested.attrset.has.a.integer.value = 100;
+    some_floaty = 29.95;
+    ## Same syntax here on these two, but they should get serialized differently:
+    # > A list may have zero or more elements, each of which can be a scalar value, an array, a group, or another list.
+    list1d = libconfig.lib.mkList [ 1 "mixed!" 5 2 ];
+    # You might also omit the mkList, as a list will be a list (in contrast to an array) by default.
+    list2d = [ 1 [ 1 1.2 "foo" ] [ "bar" 1.2 1 ] ];
+    # > An array may have zero or more elements, but the elements must all be scalar values of the same type.
+    array1d = libconfig.lib.mkArray [ 1 5 2 ];
+    array2d = [
+      (libconfig.lib.mkArray [ 1 2 ])
+      (libconfig.lib.mkArray [ 2 1 ])
+    ];
+    nasty_string = "\"\n\\\t\r^*\b\f\n\0\";'''$";
+
+    weirderTypes = {
+      _includes = [ include_file ];
+      pi = 3.141592654;
+      bigint = 9223372036854775807;
+      hex = libconfig.lib.mkHex "0x1FC3";
+      octal = libconfig.lib.mkOctal "0027";
+      float = libconfig.lib.mkFloat "1.2E-3";
+      array_of_ints = libconfig.lib.mkArray [
+        (libconfig.lib.mkOctal "0732")
+        (libconfig.lib.mkHex "0xA3")
+        1234
+      ];
+      list_of_weird_types = [
+        3.141592654
+        9223372036854775807
+        (libconfig.lib.mkHex "0x1FC3")
+        (libconfig.lib.mkOctal "0027")
+        (libconfig.lib.mkFloat "1.2E-32")
+      ];
+    };
+  };
+in
+  libconfig.generate "libconfig-test.cfg" expression
```

Comment:
What's the expected value?

This tester will be useful: https://nixos.org/manual/nixpkgs/stable/#tester-equalContents

You can compare it to an actual "hardcoded" `comprehensive.cfg` file. It might need an update if the library changes superficially its output, but that's ok. It's good to at least observe such changes.

Reply:
We validate the changes in `libconfig.generate` using the C validator which checks against the ground truth libconfig implementation. I believe requiring a bump (of sorts) of the output would only create a sort of alarm fatigue and make the existing tooling harder to use.

Reply:
I don't think we know that it would change often, or do we?
If we anticipate such a problem, we should have a program that parses it and checks the value of each field. Pick your poison ;)

Reply:
Ah, fair (though I think it's still overkill either way)

 +1 for `equalContents`

Reply:
It seems a bit superfluous to check the config text, given the test is already checking against the deserialised values.
This could also be added in a follow up PR, but it's not essential. Can we go on?

Reply:
> given the test is already checking against the deserialised values

Currently, we check whether the reference implementation can parse the generated file, not the parsed values themselves.

[`javaProperties` also compares its values to an expected value](https://github.com/nixos/nixpkgs/tree/e56990880811a451abd32515698c712788be5720/pkgs/pkgs-lib/formats/java-properties/test/default.nix), and [so do these smaller formats](https://github.com/nixos/nixpkgs/blob/e56990880811a451abd32515698c712788be5720/pkgs/pkgs-lib/tests/formats.nix#L36-L253).

Reply:
Ah, then it makes sense.